### PR TITLE
backport C++ 3d/calibration_base.cpp:5.x to calib3d/calibration_base.cpp:4.x

### PR DESCRIPTION
**Pull request by vrabaud**
This is to minimize the differences between 4.x and 5.x

To review this PR, it is best to look at the second commit which is a simpler diff.

The next steps are:
- backport stereo/stereo_geom.cpp:5.x
- backport the remaining calibration.cpp code to C++

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch

### CODE REVIEW

File: modules/calib3d/src/calibration.cpp [hunk 1]
```diff
@@ -59,460 +58,38 @@
...
-
-                for( j = 0; j < L*N; j++ )
-                    dcdb[j] = 0;
-                for( j = 0; j < L; j++ )
-                    dcdb[j*N + i2] = a[j];
-            }
-        }
-    }
+    Mat A = cvarrToMat(_A), B = cvarrToMat(_B);
+    Mat dABdA = cvarrToMat(_dABdA), dABdB = cvarrToMat(_dABdB);
```
Comment (asmorkalov) [inline modules/calib3d/src/calibration.cpp:64]: `dABdA` and `dABdB` are optional. `cvarrToMat` returns empty Mat if the pointer is nullptr. Empty Mat is not equal to `cv::NoArray()`. You have to check pointers on wrapper side and put `cv::noArray` for optional outputs.

File: modules/calib3d/src/calibration.cpp [hunk 2]
```diff
@@ -59,460 +58,38 @@
...
-    if( dt3dr1 )
-        cvZero( dt3dr1 );
+    Mat rvec1 = cvarrToMat(_rvec1), tvec1 = cvarrToMat(_tvec1),
+        rvec2 = cvarrToMat(_rvec2), tvec2 = cvarrToMat(_tvec2);
+    Mat rvec3 = cvarrToMat(_rvec3), tvec3 = cvarrToMat(_tvec3),
+        dr3dr1 = cvarrToMat(_dr3dr1), dr3dt1 = cvarrToMat(_dr3dt1),
+        dr3dr2 = cvarrToMat(_dr3dr2), dr3dt2 = cvarrToMat(_dr3dt2),
+        dt3dr1 = cvarrToMat(_dt3dr1), dt3dt1 = cvarrToMat(_dt3dt1),
+        dt3dr2 = cvarrToMat(_dt3dr2), dt3dt2 = cvarrToMat(_dt3dt2);
+    composeRT(rvec1, tvec1, rvec2, tvec2, rvec3, tvec3, dr3dr1, dr3dt1, dr3dr2, dr3dt2, dt3dr1, dt3dt1, dt3dr2, dt3dt2);
```
Comment (asmorkalov) [inline modules/calib3d/src/calibration.cpp:84]: The same for optional params.

Comment (vrabaud) [reply modules/calib3d/src/calibration.cpp]: I actually removed the function

**Comment by vrabaud** [comment]
I actually removed the C functions: they are only used internally and the C code that was calling them is not calling the C++ functions.

**Comment by vrabaud** [comment]
I just added a little fix for vertical distortion coefficients (I thought our internal failing test was flaky). This code will be removed in a future PR anyway.

**Review by asmorkalov — APPROVED** [review]

**Pull request merged**